README
NAME
    Data::Sah::Object - Object-oriented interface for Sah schemas

VERSION
    This document describes version 0.01 of Data::Sah::Object (from Perl
    distribution Data-Sah-Object), released on 2014-07-28.

SYNOPSIS
     use Data::Sah::Object; # automatically exports sah(), sahn()

     # sah() creates a normalized copy of schema
     $osch = sah("array*");
     $osch = sah(['array*', of => 'str*']);

     # sahn() assumes you're passing an already-normalized schema and it will not
     # create a copy
     my $osch = sahn([array => {req=>1, of=>'str*'}, {}]);

     say $osch->type; # -> array
     say $osch->clause('req'); # -> 1
     say $osch->clause('of', 'int'); # set clause

DESCRIPTION
    Sah works using pure data structures, but sometimes it's convenient to
    have an object-oriented interface (wrapper) for those data. This module
    provides just that.

FUNCTIONS
  sah $sch => OBJECT
    Exported by default. A shortcut for
    Data::Sah::Object::Schema->new($sch).

  sahn $sch => OBJECT
    Exported by default. A shortcut for Data::Sah::Object::Schema->new($sch,
    1).

SEE ALSO
    Sah, Data::Sah

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Data-Sah-Object>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-Data-Sah-Object>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Data-Sah-Object>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Data-Sah-Object

0.01     2014-07-28 (SHARYANTO) - Fireworks are so noisy!

         - First early release.


lib/Data/Sah/Object.pm
package Data::Sah::Object;

our $DATE = '2014-07-28'; # DATE
our $VERSION = '0.01'; # VERSION

use 5.010001;
use strict;
use warnings;

require Exporter;
our @ISA    = qw(Exporter);
our @EXPORT = qw(sah sahn);

sub sah {
    require Data::Sah::Object::Schema;
    Data::Sah::Object::Schema->new($_[0]);
}

sub sahn {
    require Data::Sah::Object::Schema;
    Data::Sah::Object::Schema->new($_[0], 1);
}

1;
# ABSTRACT: Object-oriented interface for Sah schemas

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Sah::Object - Object-oriented interface for Sah schemas

=head1 VERSION

This document describes version 0.01 of Data::Sah::Object (from Perl distribution Data-Sah-Object), released on 2014-07-28.

=head1 SYNOPSIS

 use Data::Sah::Object; # automatically exports sah(), sahn()

 # sah() creates a normalized copy of schema
 $osch = sah("array*");
 $osch = sah(['array*', of => 'str*']);

 # sahn() assumes you're passing an already-normalized schema and it will not
 # create a copy
 my $osch = sahn([array => {req=>1, of=>'str*'}, {}]);

 say $osch->type; # -> array
 say $osch->clause('req'); # -> 1
 say $osch->clause('of', 'int'); # set clause

=head1 DESCRIPTION

L<Sah> works using pure data structures, but sometimes it's convenient to have
an object-oriented interface (wrapper) for those data. This module provides just
that.

=head1 FUNCTIONS

=head2 sah $sch => OBJECT

Exported by default. A shortcut for Data::Sah::Object::Schema->new($sch).

=head2 sahn $sch => OBJECT

Exported by default. A shortcut for Data::Sah::Object::Schema->new($sch, 1).

=head1 SEE ALSO

L<Sah>, L<Data::Sah>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Data-Sah-Object>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Data-Sah-Object>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Data-Sah-Object>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Data/Sah/Object/Schema.pm
package Data::Sah::Object::Schema;

our $DATE = '2014-07-28'; # DATE
our $VERSION = '0.01'; # VERSION

use 5.010;
use strict;
use warnings;

sub new {
    my ($class, $sch, $is_normalized) = @_;
    $sch //= [undef, {}, {}];

    unless ($is_normalized) {
        require Data::Sah::Normalize;
        $sch = Data::Sah::Normalize::normalize_schema($sch);
    }

    my $obj = \$sch;
    bless $obj, $class;
}

sub type {
    my $self = shift;
    if (@_) {
        my $old = ${$self}->[0];
        ${$self}->[0] = $_[0];
        return $old;
    } else {
        return ${$self}->[0];
    }
}

sub clause {
    my $self = shift;
    my $name = shift;
    if (@_) {
        my $old = ${$self}->[1]{$name};
        ${$self}->[1]{$name} = $_[0];
        return $old;
    } else {
        return ${$self}->[1]{$name};
    }
}

sub req {
    my $self = shift;
    if (@_) {
        my $old = ${$self}->[1]{req};
        ${$self}->[1]{req} = $_[0];
        return $old;
    } else {
        return ${$self}->[1]{req};
    }
}

sub delete_clause {
    my $self = shift;
    my $name = shift;
    my $old = ${$self}->[1]{$name};
    delete ${$self}->[1]{$name};
    return $old;
}

sub as_struct {
    my $self = shift;
    ${$self};
}

1;
# ABSTRACT: Represent Sah schema

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Sah::Object::Schema - Represent Sah schema

=head1 VERSION

This document describes version 0.01 of Data::Sah::Object::Schema (from Perl distribution Data-Sah-Object), released on 2014-07-28.

=head1 METHODS

=head2 new($sch, $is_normalized) => obj

Constructor. Will create a normalized copy of C<$sch>, unless C<$is_normalized>
is true, in which case won't create a copy.

=head2 type([ $value ]) => str

Get or set type. If set, will return the old value.

=head2 clause($name[, $value]) => any

Get or set a clause. If set, will return the old value.

=head2 req([ $value ]) => bool

Shortcut for C<< $osch->clause('req'[, $value ]) >>.

=head2 delete_clause($name) => any

Delete clause named C<$name>. Return the old value.

=head2 as_struct => array

Return the schema raw data structure.

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Data-Sah-Object>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Data-Sah-Object>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Data-Sah-Object>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Object-oriented interface for Sah schemas",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Data-Sah-Object",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Data::Sah::Object",
  "PREREQ_PM" => {
    "Data::Sah::Normalize" => 0,
    "Sah" => "v0.9.30"
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Sah::Normalize" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Sah" => "v0.9.30",
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



